### DISCUSSION TITLE
Starting Remix server programmatically

### DISCUSSION THREAD

**Discussion by frontsideair** [remix-run/remix#5379] [Q&A]
Hi all,

I'm trying a new testing strategy and looking for some help. I was starting a docker-compose process in CI and testing using Playwright, but this proved to be error prone and not good for testing overall. I'm seeking a different pattern now.

What I'm trying to do is to start a Remix server with mocks in place of database, SMTP server, etc. I can then visit the page using Playwright and check if my mocks were called the right way. Here's a sample test to illustrate this point:

```ts
test("smoke test", async ({ page }) => {
  const app = await createApp({ context: { db: myDbMock, mailer: myMailerMock }, port: 3001 });

  await page.goto(app.url);
  // perform some interactions

  expect(myDbMock.data).toBe(someData);
});
```

I like this because then tests can be fast, they can run in parallel, I can start the test in whatever state that I want it in, etc.

How do I implement `createApp`? I have tried a few different approaches but none seem to work so far. I'm also using the Vercel adapter, which might be complicating things.

**Comment by sergiodxa** [comment]
I wrote how to do it here https://sergiodxa.com/articles/e2e-test-remix-with-vitest-and-puppeteer, in my case I used `execa` package to run `npm start` and set PORT to a random one so I could start it, but if you import your Express app in your tests you could do `app.listen(randomPort)` directly too, if you use Remix App Server or another thing where you don't have a way to do that then running `npm start` is simpler.

**Reply by frontsideair** [reply]
I stumbled upon your solution before and it's very approachable. But I want to be able to pass mocks into the app so I want to start my own server. I tried this but it looks like I'm gonna need to copy a lot of code and I'm getting a lot of errors, so I wanted to see if anyone tried this before.

**Reply by frontsideair** [reply]
I think I have made some headway, it was particularly difficult since while I'm working with compiled code and `require`s there's no type safety and it's hard to to figure out what fits where. I was copying code from `remix-serve`, but when using Vercel in development mode, `createRequestHandler` is already compiled into the build output.

Long story short, since `createRequestHandler` is already baked, I cannot inject my own context, so I must create a separate build or start a dev server instead.

This is mostly a note to myself or to those who might walk the same path as me. I believe this is a fruitful path and I intend to write a blog post showing the benefits of testing like this over starting a docker-compose or using a staging database.
